bgv/evaluator.go

```diff
@@ -473,106 +650,608 @@ func (eval *evaluator) mulRelin(ctIn *rlwe.Ciphertext, op1 rlwe.Operand, relin b
+			var rlk *rlwe.RelinearizationKey
+			var err error
+			if rlk, err = eval.CheckAndGetRelinearizationKey(); err != nil {
+				return fmt.Errorf("cannot Tensor: cannot Relinearize: %w", err)
 			}
 
-			tmpCt := &rlwe.Ciphertext{Value: []*ring.Poly{eval.BuffQP[1].Q, eval.BuffQP[2].Q}}
+			tmpCt := &rlwe.Ciphertext{}
+			tmpCt.Value = []ring.Poly{eval.BuffQP[1].Q, eval.BuffQP[2].Q}
+			tmpCt.MetaData = &rlwe.MetaData{}
 			tmpCt.IsNTT = true
 
-			eval.GadgetProduct(level, c2, eval.Rlk.Keys[0].GadgetCiphertext, tmpCt)
+			eval.GadgetProduct(level, c2, &rlk.GadgetCiphertext, tmpCt)
 
-			ringQ.Add(ctOut.Value[0], tmpCt.Value[0], ctOut.Value[0])
-			ringQ.Add(ctOut.Value[1], tmpCt.Value[1], ctOut.Value[1])
+			ringQ.Add(opOut.Value[0], tmpCt.Value[0], opOut.Value[0])
+			ringQ.Add(opOut.Value[1], tmpCt.Value[1], opOut.Value[1])
 		}
 
 		// Case Plaintext (x) Ciphertext or Ciphertext (x) Plaintext
 	} else {
 
-		if ctOut.Degree() < ctIn.Degree() {
-			ctOut.Resize(ctIn.Degree(), level)
-		}
+		opOut.Resize(op0.Degree(), level)
 
 		c00 := eval.buffQ[0]
 
-		ringQ.MForm(op1.El().Value[0], c00)
-		ringQ.MulScalar(c00, eval.params.T(), c00)
-		for i := range ctOut.Value {
-			ringQ.MulCoeffsMontgomery(ctIn.Value[i], c00, ctOut.Value[i])
+		// Multiply by T * 2^{64} * 2^{64} -> result multipled by T and switched in the Montgomery domain
+		ringQ.MulRNSScalarMontgomery(op1.El().Value[0], eval.tMontgomery, c00)
+		for i := range opOut.Value {
+			ringQ.MulCoeffsMontgomery(op0.Value[i], c00, opOut.Value[i])
+		}
+	}
+
+	return
+}
+
+// MulScaleInvariant multiplies op0 with op1 without relinearization and using scale invariant tensoring (BFV-style), and returns the result in opOut.
+// This tensoring increases the noise by a constant factor regardless of the current noise, thus no rescaling is required with subsequent multiplications if they are
+// performed with the invariant tensoring procedure. Rescaling can still be useful to reduce the size of the ciphertext, once the noise is higher than the prime
+// that will be used for the rescaling or to ensure that the noise is minimal before using the regular tensoring.
+// The procedure will return an error if either op0.Degree or op1.Degree > 1.
+// The procedure will return an error if the evaluator was not created with an relinearization key.
+//
+// inputs:
+//   - op0: an *rlwe.Ciphertext
+//   - op1:
+//   - rlwe.ElementInterface[ring.Poly]
+//   - *big.Int, uint64, int64, int
+//   - []uint64 or []int64 (of size at most N where N is the smallest integer satisfying PlaintextModulus = 1 mod 2N)
+//   - opOut: an *rlwe.Ciphertext
+//
+// If op1 is an rlwe.ElementInterface[ring.Poly]:
+//   - the level of opOut will be updated to min(op0.Level(), op1.Level())
+//   - the scale of opOut will be to op0.Scale * op1.Scale * (-Q mod T)^{-1} mod T
+func (eval Evaluator) MulScaleInvariant(op0 *rlwe.Ciphertext, op1 rlwe.Operand, opOut *rlwe.Ciphertext) (err error) {
```

Comment:
I would suggest adding a clarification that explains when MulScaleInvariant should be used over Mul and why

Reply:
The doc of the method specifies this is the BFV-style tensoring. This then to the user's choice to chose between the two tensoring based on the application. There is no definite answer, it depends on the circuit and the current noise level.